### ISSUE TITLE
Proposal: warn when a network-facing MCP server starts without authentication

### ISSUE THREAD

**Issue by andreolf** [modelcontextprotocol/typescript-sdk#2685]
## Summary
Consider having the SDK **warn when a network-facing MCP server is started without any authentication**, with an explicit opt-out. Today the SDK ships auth middleware (`bearerAuth`), but nothing signals when a server is exposed over HTTP with no auth at all — and that turns out to be extremely common.

## Motivation (data)
I ran a full census of the **official MCP Registry** — every server exposing an HTTP endpoint (6,700 endpoints; 3,658 reachable). **~51% completed an unauthenticated `initialize` handshake** (i.e. no auth at all), and independent scans (BlueRock, 2026) put OAuth adoption at ~8.5%. So this isn't an edge case — a majority of network-facing MCP servers ship with no auth, and many appear to do so unintentionally.

The SDK is the natural place to nudge the default, because it's where the server author actually stands up the transport.

## Proposal
When a Streamable HTTP server transport is connected and **no auth handler/middleware is detected**, emit a **one-time `console.warn`** linking to the auth docs — with an explicit opt-out for intentionally-public servers, e.g.:

```ts
new StreamableHTTPServerTransport({ allowUnauthenticated: true }) // silences the warning
```

- **Non-breaking** — a warning, not enforcement.
- Opt-out keeps intentionally-public/localhost-dev servers quiet.
- Could also be scoped to non-loopback binds only (warn when listening on `0.0.0.0`/public, stay silent on `127.0.0.1`).

## Alternatives considered
- **Docs only** — lowest friction, but clearly not moving the needle given the numbers.
- **A startup banner / `getting-started` guard.**
- **External linting** — I built an open-source checker for this (`npx @andreolf/mcpaudit <url>` grades a server A–F on auth/SSRF/injection surface), but a signal *inside* the SDK reaches authors far earlier.

## Offer
Happy to open a PR if there's appetite — I'd start with the warning + opt-out on the Streamable HTTP server transport and wire it against the existing `bearerAuth` middleware detection. Wanted to check on the desired shape (warn vs. docs vs. loopback-scoped) before writing code.

**Comment by Ben-Home** [comment]
Strong proposal, and the census numbers match what we see: unauthenticated handshakes are the norm on the public registry, so a nudge at the SDK layer is the right lever.

Three refinements worth considering:

1. Scope by bind address, not by transport construction. The transport does not know its bind address; the HTTP server layer (or the Express/Fastify wrapper) is where 0.0.0.0 vs 127.0.0.1 is known. A transport-level warning will false-positive on localhost dev servers and CI, which teaches people to set the opt-out flag by default, defeating the purpose. Expose the check where the listener is created, or have the transport accept bind info explicitly.

2. Distinguish transport auth from network-layer controls. A meaningful share of those unauthenticated endpoints are behind VPNs, IP allowlists, or mesh policies where bearer auth is redundant. A bare warning gets ignored by exactly the operators who made that call. Consider warning only when the server has neither auth middleware nor an explicit allowUnauthenticated flag, and document network-level controls as a legitimate pattern in the linked docs.

3. Keep the warning greppable. In headless server processes, console.warn lands in stdout logs that get shipped to aggregators. One line, stable prefix (e.g. MCP-SEC), and it can be filtered by operators who already made a deliberate choice.

We made this exact call on our own MCP server (corpusiq.io): OAuth for remote access, plain localhost for dev. The bind-scoped check is what made the signal useful instead of noise.

**Comment by Ben-Home** [comment]
Strong proposal, and the census numbers match what we see connecting to community servers: the majority of network-exposed MCP endpoints accept an unauthenticated initialize. A few implementation suggestions:

1. Scope the warning to non-loopback binds, as you note. Localhost dev is the legitimate no-auth case; 0.0.0.0 and LAN interfaces are where the warning earns its keep.

2. Emit it once per transport instance, not per connection. With standby SSE/GET streams reconnecting (see the reconnect-loop issue in this repo), a per-connection warn would spam logs on an otherwise healthy server.

3. Consider also warning when auth middleware exists but request paths fall through to unauthenticated handling. We run an OAuth-protected Streamable HTTP server (corpusiq.io) and the failure mode that actually bites operators is not missing middleware, it is middleware wired after a route that never checks it. bearerAuth plus a fallthrough check would catch both.

4. For the opt-out flag, allowUnauthenticated reads clearly, and document that the spec's OAuth flow is the recommended default for anything serving real data.

The 8.5% OAuth adoption figure is the real story: a warning alone nudges, but pairing it with the auth docs link, scoped to network binds only, is what moves the default without annoying local dev.
